### PR TITLE
Add functionality to purchase shares through the UI

### PR DESCRIPTION
**Pull request by Artem-Semenov-dev**
This PR adds functionality to purchase shares through the UI.
The view of the purchase process:
Screen Recording 2023-06-08 at 16 12 33

### CODE REVIEW

File: client/src/main/kotlin/io/spine/examples/shareaware/client/MarketPage.kt [hunk 1]
```diff
@@ -81,6 +109,184 @@ public class MarketPageModel(client: DesktopClient) {
...
+     * whether it is present or not.
+     */
+    public fun purchaseResultMessageShown(): StateFlow<Boolean> {
+        return purchaseResultMessageShown
+    }
+
+    /**
+     * Returns the current state of the message that signals about the purchase operation result.
+     *
+     * It needs to be shown to the user to inform about the result of the purchase operation.
```
Comment (Vladyslav-Kuksiuk) [inline client/src/main/kotlin/io/spine/examples/shareaware/client/MarketPage.kt:172]: I think it is redundant.

File: client/src/main/kotlin/io/spine/examples/shareaware/client/MarketPage.kt [hunk 2]
```diff
@@ -81,6 +109,184 @@ public class MarketPageModel(client: DesktopClient) {
...
+     */
+    public fun purchaseResultMessage(): StateFlow<String> {
+        return purchaseResultMessage
+    }
+
+    /**
+     * Returns the current state of the purchase operation, whether it failed or not.
+     */
+    public fun purchaseFailed(): StateFlow<Boolean> {
+        return purchaseFailed
```
Comment (Vladyslav-Kuksiuk) [inline client/src/main/kotlin/io/spine/examples/shareaware/client/MarketPage.kt:182]: I think it's better to rename it to `isPurchaseFailed`.

File: client/src/main/kotlin/io/spine/examples/shareaware/client/MarketPage.kt [hunk 3]
```diff
@@ -81,6 +109,184 @@ public class MarketPageModel(client: DesktopClient) {
...
+        purchaseFailed.value = false
+        purchaseResultMessage.value = ""
+    }
+
+    /**
+     * Sends the `PurchaseShares` command to the server.
+     */
+    public fun purchaseShares() {
+        val share = shareToPurchase.value
+        Preconditions.checkNotNull(share)
```
Comment (Vladyslav-Kuksiuk) [inline client/src/main/kotlin/io/spine/examples/shareaware/client/MarketPage.kt:220]: Let's import `checkNotNull` directly.

**Review by Vladyslav-Kuksiuk — COMMENTED** [review]
@Artem-Semenov-dev mostly LGTM, please see my comments.

File: client/src/main/kotlin/io/spine/examples/shareaware/client/MarketPage.kt [hunk 4]
```diff
@@ -81,6 +109,181 @@ public class MarketPageModel(client: DesktopClient) {
     public fun shares(): StateFlow<AvailableMarketShares?> {
         return sharesSubscriptions.state()
     }
+
+    /**
+     * Sets the page to "purchase" state.
+     *
+     * Page state when the user wants to purchase shares.
```
Comment (armiol) [inline client/src/main/kotlin/io/spine/examples/shareaware/client/MarketPage.kt:116]: That's not English. Please rephrase.

File: client/src/main/kotlin/io/spine/examples/shareaware/client/MarketPage.kt [hunk 5]
```diff
@@ -81,6 +109,181 @@ public class MarketPageModel(client: DesktopClient) {
...
+     *
+     * @param share the share to purchase
+     */
+    public fun toPurchaseState(share: Share) {
```
Comment (armiol) [inline client/src/main/kotlin/io/spine/examples/shareaware/client/MarketPage.kt:120]: Let's speak in terms of domain. E.g. `initiatePurchase` and `complete...`, `cancel...`, etc.

File: client/src/main/kotlin/io/spine/examples/shareaware/client/MarketPage.kt [hunk 6]
```diff
@@ -81,6 +109,181 @@ public class MarketPageModel(client: DesktopClient) {
...
+     */
+    public fun toDefaultState() {
+        purchaseState.value = false
+        quantityToPurchase.value = 0
+    }
+
+    /**
+     * Returns the "purchase" state of the page.
+     */
+    public fun purchaseState(): StateFlow<Boolean> {
```
Comment (armiol) [inline client/src/main/kotlin/io/spine/examples/shareaware/client/MarketPage.kt:136]: `purchaseInProgress`?

File: client/src/main/kotlin/io/spine/examples/shareaware/client/MarketPage.kt [hunk 7]
```diff
@@ -54,23 +63,42 @@ import androidx.compose.ui.platform.LocalDensity
...
 import kotlinx.coroutines.withContext
 
 /**
  * UI model for the `MarketPage`.
  */
-public class MarketPageModel(client: DesktopClient) {
+public class MarketPageModel(private val client: DesktopClient) {
     private val sharesSubscriptions: EntitySubscription<AvailableMarketShares> =
         EntitySubscription(AvailableMarketShares::class.java, client, MarketProcess.ID)
+    private val purchaseState: MutableStateFlow<Boolean> = MutableStateFlow(false)
```
Comment (armiol) [inline client/src/main/kotlin/io/spine/examples/shareaware/client/MarketPage.kt:96]: I think we should have at least a separate type for this operation. 

Right now it's written in a procedural style.

**Review by armiol — CHANGES_REQUESTED** [review]

**Review by armiol — APPROVED** [review]
Still with a number of both UI and implementation aspects to improve. But I think it's good enough for now.

Please feel free to merge, and let's start working on UI improvements.

**Pull request merged**